### PR TITLE
Threading support for legacy loading of checkpoints

### PR DESCRIPTION
**Pull request by krshrimali**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/12557.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: pytorch_lightning/utilities/migration.py [hunk 1]
```diff
@@ -34,7 +35,12 @@ class pl_legacy_patch:
             torch.load("path/to/legacy/checkpoint.ckpt")
     """
 
+    def __init__(self) -> None:
+        # Create a lock to ensure no race condition with deleting sys modules
+        self.lock = threading.Lock()
```
Comment (Borda) [inline pytorch_lightning/utilities/migration.py:40]: Rather protected?

Comment (krshrimali) [reply pytorch_lightning/utilities/migration.py]: That's a good suggestion, @Borda - thanks! I've made it protected now.

**Comment by krshrimali** [comment]
Update: working on adding tests for this PR, and then it should be ready for review. :)

File: pytorch_lightning/utilities/migration.py [hunk 2]
```diff
@@ -34,7 +35,12 @@ class pl_legacy_patch:
             torch.load("path/to/legacy/checkpoint.ckpt")
     """
 
+    def __init__(self) -> None:
+        # Create a lock to ensure no race condition with deleting sys modules
+        self._lock = threading.Lock()
```
[the remark this answers is not in the document]
Comment (krshrimali) [reply pytorch_lightning/utilities/migration.py]: Thanks, @awaelchli - you were absolutely right!

For others, post offline discussion with Adrian, and thanks to him for helping out with an explanation, a global lock variable is a correct solution here. This has been fixed now in this PR, and a test has been added.

**Review by carmocca — APPROVED** [review]

**Review by otaj — APPROVED** [review]

**Review by otaj — APPROVED** [review]

**Pull request merged**